Wrapping up documentation and some mild testing

Hector-Lantern opened the pull request:
I tested the intercept command manually, wrote a short blurb in the readme on that. Validate works and has a test case that reads the strategies.txt and validates each. Now the user will have more strategies to choose from. I would like a link to any documentation explaining the strategies so the user can make an informed decision.

Cut out some comments and print statements I used for debugging.

I didn't get PCAP tested, it needs some work and I don't fully understand it yet. We should make a new ticket for that.

Hector-Lantern committed:
Added to readme and cleaned up various unneccesary print statements and comments

hwh33 commented on README.md:
```diff
@@ -39,10 +38,37 @@ You can find a list of available interfaces using
 
+## How to test
+
+You can test or rather verify that the intercept mode is running by simply running it with a validated strategy. The output should show that packets are being rerouted. You can monitor your adapter in [WireShark](https://www.wireshark.org/) and you should notice a huge uptick in packets being sent from your adapter.
```
The only strategy I've been able to run is the one in s.txt.  Which other strategies did you test with?  Also many strategies generate error statements, but the reference is malformed and the error is `nil`.  Here's an example using `[TCP:flags:S]-duplicate(tamper{TCP:flags:replace:SA},)-| \/`, which is pulled straight from strategies.txt:

```
.\geneva-cli.exe intercept -interface "vEthernet (Virtual Switch - WiFi Adapter)" -strategyFile .\from-strategies.txt
I: 14:34:17 outbound \/ inbound {[] ["[TCP:flags:S]-duplicate(tamper{TCP:flags:replace:SA}(send,),)-|"]}
I: 14:34:17 starting up
I: 14:34:17 starting statistics thread
I: 14:34:17 intercepting traffic on vEthernet (Virtual Switch - WiFi Adapter) (idx 11)
I: 14:34:17 using filter "ip and !loopback and ifIdx == 11"
I: 14:34:17 opening handle to WinDivert
E: 14:34:22 failed to send packet after error applying strategy: %!w(<nil>)
I: 14:34:22 intercepted/injected packets: 67/67 inbound, 87/88 outbound; 1 errors
E: 14:34:22 failed to send packet after error applying strategy: %!w(<nil>)
... same line 14 more times ...
I: 14:34:26 stopping all threads
I: 14:34:26 closing WinDivert handle
I: 14:34:26 intercepted/injected packets: 820/820 inbound, 756/772 outbound; 16 errors
```

I was also unable to verify that the strategy specified in s.txt was actually applied as expected.  How did you verify this?

hwh33 replied:
The program also panics if the strategy file does not exist.

Hector-Lantern replied:
So the code expects one strategy passed to it in a file, strategies.txt is supposed to be a reference of known useful strategies. I got it from the main Geneva repository. I should make clear in the readme that passing strategies.txt to intercept is a bad idea.

For the panic, I will update the code to check for the file first and output an error.

hwh33 commented on README.md:
```diff
@@ -39,10 +38,37 @@ You can find a list of available interfaces using
+
+You can test or rather verify that the intercept mode is running by simply running it with a validated strategy. The output should show that packets are being rerouted. You can monitor your adapter in [WireShark](https://www.wireshark.org/) and you should notice a huge uptick in packets being sent from your adapter.
+
+You can run `go test` to run some unit tests, currently there is only tests for validating strategies.txt
```
There are no tests...
```
go test
?       github.com/getlantern/geneva-cli        [no test files]
```

Hector-Lantern replied:
I forgot to commit it

hwh33 commented on README.md:
````diff
@@ -39,10 +38,37 @@ You can find a list of available interfaces using
+
+### Start
+
+```Start-Service -Name "Geneva-CLI"```
````
This did not work for me.  I ran all instructions as specified in the README in this branch.  Did I screw something up?
```
> New-Service -Name "Geneva-CLI" -DisplayName "WinGeneva" -Description "Geneva CLI for Windows" -StartupType Manual -BinaryPathName "C:\Users\Harry\Code\Lantern\geneva-cli\geneva-cli.exe intercept -interface 'vEthernet (Virtual Switch - WiFi Adapter)' -strategyFile C:\Users\Harry\Code\Lantern\geneva-cli\s.txt"

Status   Name               DisplayName
------   ----               -----------
Stopped  Geneva-CLI         WinGeneva


> Start-Service -Name "Geneva-CLI"
Start-Service : Service 'WinGeneva (Geneva-CLI)' cannot be started due to the following error:
Cannot start service Geneva-CLI on computer '.'.
At line:1 char:1
+ Start-Service -Name "Geneva-CLI"
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OpenError: (System.ServiceProcess.ServiceController:ServiceController)
    [Start-Service], ServiceCommandException
    + FullyQualifiedErrorId : CouldNotStartService,Microsoft.PowerShell.Commands.StartServiceCommand
```

Crosse replied:
The System or Application log may have more info?

Hector-Lantern replied:
Uhh double check that PS is in administrator mode?

hwh33 commented on README.md:
````diff
@@ -39,10 +38,37 @@ You can find a list of available interfaces using
+
+### Install Command, in manual mode
+
+```New-Service -Name "Geneva-CLI" -DisplayName "WinGeneva" -Description "Geneva CLI for Windows" -StartupType Manual -BinaryPathName "Path-To-Geneva.exe --args"```
````
Bit of a nit, but I don't think the service itself should be called "Geneva-CLI"; CLI refers to the command-line program (which I understand the service is utilizing).  You could just call it "Geneva".  Same goes for the description.

hwh33 reviewed (commented):
> Validate works and has a test case that reads the strategies.txt and validates each.

(1) Validate does not appear to work for me:
```
> cat .\from-strategies.txt
[TCP:flags:S]-duplicate(tamper{TCP:flags:replace:SA},)-| \/
> .\geneva-cli.exe validate .\from-strategies.txt
invalid strategy: expected token "[" not found at position 0 ☺☺☺☺☺☺☺☺☺☺☺☺☺☺☺☺
> cat .\s.txt
[TCP:flags:PA]-fragment{tcp:-1:True}-| \/
> .\geneva-cli.exe validate .\s.txt
invalid strategy: expected token "[" not found at position 0 ☺☺☺☺☺☺☺☺☺☺☺☺☺☺☺☺
```

(2) What do you mean when you talk about a test case for each strategy in strategies.txt?  Is there some test code which has not been pushed?

> I would like a link to any documentation explaining the strategies so the user can make an informed decision.

Where did you get the strategies.txt file from?

> I didn't get PCAP tested, it needs some work and I don't fully understand it yet. We should make a new ticket for that.

Could you open a new ticket for this?  Could you link the ticket to [this epic](https://github.com/getlantern/engineering/issues/373)?  The ticket should cover testing and fixing all remaining sub-commands of the Geneva CLI tool; please list in the ticket each of the sub-commands which still need testing.

Hector-Lantern committed:
Forgot to commit validate_test.go

Crosse reviewed (commented):
@hwh33 asked:
> Where did you get the strategies.txt file from?

From the geneva library repo: https://github.com/getlantern/geneva/blob/main/strategies.txt. I, in turn, copied them from [the upstream repo](https://github.com/Kkevsterrr/geneva/blob/master/strategies.md) and from the [paper itself](https://geneva.cs.umd.edu/papers/2019-ccs-geneva.pdf) (pg 2207). There might be other sources I've forgotten, but tl;dr they all came from the Geneva folks.

> (1) Validate does not appear to work for me[...]

Hm, it does for me. I added a bogus strategy to `strategies.txt` and ran the following snippet. The only strategy that failed to validate was my bogus one.

```
$ while read -r strat; do \
    out=$(./geneva-cli validate "$strat" 2>&1 | tr -d '\0'); \
    if [[ "$out" =~ invalid ]]; then \
      printf 'failed to validate "%s": %s\n' "$strat" "$out"; \
    fi; \
done < <(awk '/.+/ { print }' strategies.txt)

failed to validate "wtf no": invalid strategy: expected token "[" not found at position 0
```

EDIT: Ohhhh, I see the issue. The `validate` subcommand takes a _strategy_ as input, not a file. :slightly_smiling_face: It might be good to augment the subcommand help text to call that out.

@Hector-Lantern said:
> I would like a link to any documentation explaining the strategies so the user can make an informed decision.

You can check out the following:
- the [original paper](https://geneva.cs.umd.edu/papers/2019-ccs-geneva.pdf), &sect; 3.2 specifically;
- The [upstream repo](https://github.com/Kkevsterrr/geneva/tree/master#Censorship-Evasion-Strategies);
- my own rewriting of the above in our [Geneva repo](https://github.com/getlantern/geneva#strategies-forests-and-action-trees).

---

I didn't catch it before, but the `list-adapters` command breaks non-Windows builds:
```

$ go env | grep -E 'GO(ARCH|OS)' ; go build
GOARCH='amd64'
GOOS='linux'
# github.com/getlantern/geneva-cli
./main.go:85:11: undefined: listAdaptersWrapper
```

It would be best to handle it like [the `dot` subcommand on Windows,](https://github.com/getlantern/geneva-cli/blob/main/dot_windows.go) where I just stubbed it out and said "no `dot` for you!" Another option is to simply move the subcommand to `init()` in `intercept_windows.go`.

Also, line 38 of `dot.go` needs a comma after the closing brace. Building failed on Linux because I missed that when building on Windows.

The following diff should fix both of these issues:

```diff
diff --git a/dot.go b/dot.go
index 0bffd12..e5ff96d 100644
--- a/dot.go
+++ b/dot.go
@@ -35,7 +35,7 @@ func init() {
 			},
 			ArgsUsage: "STRATEGY",
 			Action:    dot,
-		}
+		},
 	)
 }
 
diff --git a/intercept_windows.go b/intercept_windows.go
index 5d0c0ce..bb37d13 100644
--- a/intercept_windows.go
+++ b/intercept_windows.go
@@ -24,6 +24,13 @@ var (
 )
 
 func init() {
+	app.Commands = append(app.Commands, &cli.Command{
+		Name:   "list-adapters",
+		Usage:  "Lists the available adapters",
+		Action: listAdaptersWrapper,
+		Flags:  []cli.Flag{},
+	})
+
 	dll := windows.NewLazySystemDLL("kernel32.dll")
 	if err := dll.Load(); err != nil {
 		panic(fmt.Errorf("error loading kernel32.dll: %v", err))
diff --git a/main.go b/main.go
index 55275f5..e78b951 100644
--- a/main.go
+++ b/main.go
@@ -79,13 +79,6 @@ func init() {
 		return nil
 	}
 
-	app.Commands = append(app.Commands, &cli.Command{
-		Name:   "list-adapters",
-		Usage:  "Lists the available adapters",
-		Action: listAdaptersWrapper,
-		Flags:  []cli.Flag{},
-	})
-
 	app.Commands = append(app.Commands, &cli.Command{
 		Name:   "saved-command",
 		Usage:  "Runs commands from config file",
```

Hector-Lantern commented:
> I didn't catch it before, but the `list-adapters` command breaks non-Windows builds:

I can fix that

Hector-Lantern committed:
Added more testing for the validate command
Added command-loader.go to manage loading saved saved_commands
Updated intercept.go with some service related changes

Hector-Lantern committed:
Lots of small changes to error messages and removing unecessary code
Moved code to read saved_commands file to its own file and added unit tests
Changed list-adapters command to only be callable on windows, other platforms will give an error message
Reworked how running as a service handles parameters
Updated README.md

Hector-Lantern committed:
Changed the command cli saved-command to load-command
Made savedCommands struct use a map instead of an array
Updated saved_commands.json to use a dictionary with the command name as keys

Hector-Lantern committed:
Fixed bug with saving a command from the intercept command

Hector-Lantern committed:
Moved exclusivity check to after service check in intercept.go \nfixed issue in intercept_windows.go printing the wrong error

Hector-Lantern committed:
Fixed `go test` not working correctly
Added testdata folder with copies of strategy.txt and s.txt for testing
Changed relative paths to absolute paths that use the execPath to find saved_Commands.json

Hector-Lantern committed:
Updated tests to use testdata and filepath instead of raw strings

Hector-Lantern committed:
Fixed bug in pcap related to outputting new packets
Updated pcap command to use the same strategy and strategyFile options as intercept for consistency
Updated intercept command to pass along a start type to the service

Hector-Lantern committed:
A few minor changes, updated go.mod

Hector-Lantern committed:
Updated README.md and added a python script to help checking strategies
go.mod and go.sum were updated to use the newest version of geneva go
minor updated to intercept.go
run_pcap.go had some output changes to better debug with

Crosse commented on README.md:
```diff
@@ -39,18 +39,34 @@ Then you can run the program using
 ## How to test
 
-You can test or rather verify that the intercept mode is running by simply running it with a validated strategy. The output should show that packets are being rerouted. You can monitor your adapter in [WireShark](https://www.wireshark.org/) and you should notice a huge uptick in packets being sent from your adapter.
+You can test or rather verify that the intercept mode is running by simply running it with a validated strategy. The output should show that packets are being rerouted. You can monitor your adapter in [WireShark](https://www.Wireshark.org/) and you should notice a huge uptick in packets being sent from your adapter.
```
```suggestion
You can test or rather verify that the intercept mode is running by simply running it with a validated strategy. The output should show that packets are being rerouted. You can monitor your adapter in [Wireshark](https://www.wireshark.org/) and you should notice a huge uptick in packets being sent from your adapter.
```

Hector-Lantern replied:
So funnily this happened because I did a find and replace to capitalize the 'W' elsewhere

Crosse commented on README.md:
````diff
@@ -39,18 +39,34 @@ Then you can run the program using
 
 ```[TCP:flags:PA]-fragment{tcp:5:True}-| \/```
 
+If you are correlating the include pcap output with Wireshark, the Wireshark numbers packets starting at 1 and not 0
````
```suggestion
If you are correlating the include pcap output with Wireshark, be aware that Wireshark numbers packets start at 1 and not 0.
```

Crosse commented on README.md:
```````diff
@@ -71,15 +87,17 @@ While running the service you can check [Event Viewer](https://learn.microsoft.c
 ```.\geneva-cli.exe intercept --service stop```
 
-### Uninstall
+### Uninstall``````
```````
```suggestion
### Uninstall
```

Crosse commented on bulk_pcap.py:
```diff
@@ -0,0 +1,56 @@
+import os
```
```suggestion
#!/usr/bin/env python3

import os
```

Hector-Lantern replied:
I will batch this and the rest of the suggestions once @hwh33 looks over the changes.

Hector-Lantern committed:
Removed invalid test strategies from testdata/strategies.txt

hwh33 commented on README.md:
```diff
@@ -27,22 +27,77 @@ Rename the x64 dll to WinDivert64.dll
+
+`.\geneva-cli.exe intercept --interface <interface-name> -strategy "[TCP:flags:PA]-fragment{tcp:-1:True}-| \/"`
+
+[TCP:flags:PA]-fragment{tcp:8:True}(,fragment{tcp:4:True})-| \/
```
What is this line?

Hector-Lantern replied:
A mistake, I had the wrong window open at some point trying to copy a strategy to console

hwh33 commented on bulk_pcap.py:
```diff
@@ -0,0 +1,56 @@
+            print(f'{key}: {value}')
+
+if __name__ == '__main__':
+    main()
```
Could you add a bit of documentation?  Even just a usage string would suffice.  Something to tell readers what the script is supposed to do and what arguments are expected.  You could output the usage string when no arguments are provided.

hwh33 commented on command-loader.go:
```diff
@@ -0,0 +1,85 @@
+	err = json.Unmarshal(data, &savedComs)
+
+	if err != nil {
+		fmt.Println(err)
```
Why are you printing these errors out?  Is this just leftover debugging code?

Hector-Lantern replied:
Leftover debug code, it will be removed

hwh33 commented on command-loader.go:
```diff
@@ -0,0 +1,85 @@
+		return com, nil
+	}
+
+	return *new(Command), errors.New("no command found")
```
Just FYI, the idiomatic way to say this is `Command{}` - but this works too.

hwh33 commented on command-loader_test.go:
```diff
@@ -0,0 +1,75 @@
+
+	if _, err := os.Stat(saveFile); errors.Is(err, os.ErrNotExist) {
+		os.Remove(saveFile)
+	}
```
This removes the file only if it doesn't exist.  I'm guessing that's backwards to what you were going for.

hwh33 commented on command-loader_test.go:
```diff
@@ -0,0 +1,75 @@
+	sc.add(two)
+
+	err := sc.save(saveFile)
+	assert.Nil(t, err)
```
This works, but FYI `assert.NoError(t, err)` will provide a slightly more helpful output message.

Hector-Lantern replied:
Good to know, I will update

hwh33 commented on command-loader_test.go:
```diff
@@ -0,0 +1,75 @@
+	assert.Nil(t, err)
+
+	if _, err := os.Stat(saveFile); errors.Is(err, os.ErrNotExist) {
+		assert.Nil(t, err)
```
If you hit this line, you know the error cannot be nil (because the if statement condition requires the error to be equal to or caused by `os.ErrNotExist`).  What is this trying to check for?  That the file exists?  You could just do:
```
_, err := os.Stat(saveFile)
assert.NoError(t, err)
```

hwh33 commented on command-loader_test.go:
```diff
@@ -0,0 +1,75 @@
+	if _, err := os.Stat(saveFile); errors.Is(err, os.ErrNotExist) {
+		assert.Nil(t, err)
+	}
+	assert.Equal(t, 2, len(sc.Items))
```
Shouldn't we check the contents of the file?  I guess that's tested in `TestLoadCommands`?

hwh33 commented on command-loader_test.go:
```diff
@@ -0,0 +1,75 @@
+
+	if _, err := os.Stat(saveFile); errors.Is(err, os.ErrNotExist) {
+		os.Remove(saveFile)
+	}
```
Same as above.

hwh33 commented on main.go:
```diff
@@ -117,3 +74,13 @@ func main() {
+func getExecPath() string {
+	ex, err := os.Executable()
+	if err != nil {
+		fmt.Println(err)
+		return ""
```
This function should just return an error.  As it stands, this will lead to more confusing errors downstream (since filepaths will not be as expected).  Neither use of `getExecPath` even checks the output against the empty string.

hwh33 commented on run_pcap.go:
```diff
@@ -101,12 +109,38 @@ func NewFlow(pkt gopacket.Packet) *Flow {
-	s, err := geneva.NewStrategy(strat)
+	strat, err := geneva.NewStrategy(stratString)
 	if err != nil {
+		fmt.Printf("invalid strategy: %s", err)
```
What's with this random print statement?

Hector-Lantern replied:
It was so when I run it from the python script I could see what the error was

hwh33 commented on validate.go:
```diff
@@ -12,19 +15,106 @@ func init() {
+func validateFromCLI(c *cli.Context) error {
+
+	// command-line options can override some of flashlight's config
+	if c.String("strategy") != "" && c.String("strategyFile") != "" && c.String("bulkFile") != "" {
+		return cli.Exit("Only one of -strategy, -strategyFile, or -bulkFile should be used", 1)
+	}
```
This doesn't guard against cases where 2 of the 3 are non-empty.

hwh33 commented on validate.go:
```diff
@@ -12,19 +15,106 @@ func init() {
+			return cli.Exit(fmt.Sprintf("cannot open %s: %v", s, err), 1)
+		}
+		s = string(in)
+		fmt.Println(s)
```
?

Hector-Lantern replied:
Old debug line that I forgot about

hwh33 reviewed (commented):
In addition to the comments:

I can't get the service to run.  Am I doing something wrong?  It fails whether I provide a strategy via the command-line or via file.
```
PS C:\Users\Harry\Code\Lantern\geneva-cli> go build
PS C:\Users\Harry\Code\Lantern\geneva-cli> .\geneva-cli.exe intercept --service install
PS C:\Users\Harry\Code\Lantern\geneva-cli> .\geneva-cli.exe intercept --service start --strategy "[TCP:flags:PA]-fragment{tcp:5:True}-| \/"
E: 21:19:36 Valid actions: [start stop restart install uninstall]
Failed to start Geneva proxy: The service did not respond to the start or control request in a timely fashion.
PS C:\Users\Harry\Code\Lantern\geneva-cli> cat .\s.txt
[TCP:flags:PA]-fragment{tcp:5:True}-|\/
PS C:\Users\Harry\Code\Lantern\geneva-cli> .\geneva-cli.exe intercept --service start --strategyFile "C:\Users\Harry\Code\Lantern\geneva-cli\s.txt"
E: 21:22:28 Valid actions: [start stop restart install uninstall]
Failed to start Geneva proxy: The service did not respond to the start or control request in a timely fashion.
```

For both runs, EventViewer shows the same log:
```
cannot open s.txt: open s.txt: The system cannot find the file specified.
```

Seems like the following default arguments, defined in the intercept command, are not being overridden as expected:
```
svcConfig := service.Config{
		Name:        "geneva-proxy",
		DisplayName: "Geneva proxy",
		Description: "Geneva proxy",
		Arguments:   []string{"intercept", "-strategyFile", "s.txt"},
		Option:      options,
	}
```

Hector-Lantern committed:
Tidied up unnecessary console outputs
Added a verbose output option to run-pcap and intercept
Fixed intercept's service not applying parameters correctly
Fixed validate's flag checking to look for mutual exclusivity correctly
Updated README.md
Updated bulk_pcap.py with help message
Updated commander-loader_test to check if files exist correctly and more testing of contents

Hector-Lantern committed:
Tidied up unnecessary console outputs
Added a verbose output option to run-pcap and intercept
Fixed intercept's service not applying parameters correctly
Fixed validate's flag checking to look for mutual exclusivity correctly
Updated README.md
Updated bulk_pcap.py with help message
Updated commander-loader_test to check if files exist correctly and more testing of contents
Updated getExec to return an error

Hector-Lantern committed:
Merge branch 'testing' of github.com:getlantern/geneva-cli into testing

Hector-Lantern committed:
Updated intercept with more verbose output
Updated go.mod with new version of geneva library

Hector-Lantern committed:
updated README.md

Hector-Lantern committed:
removed unnecessary outputs
